Documentation/design-docs/object-stack-allocation.md

```diff
@@ -0,0 +1,190 @@
+# Object Stack Allocation
+
+This document describes work to enable object stack allocation in .NET Core.
+
+## Motivation
+
+In .NET instances of object types are allocated on the garbage-collected heap.
+Such allocations have performance overhead at garbage collection time. The allocator also has to ensure that the memory is fully zero-initialized.
+If the lifetime of an object is bounded by the lifetime of the allocating method, the allocation
+may be moved to the stack. The benefits of this optimization:
+
+* The pressure on the garbage collector is reduced because the GC heap becomes smaller.
+* Object field accesses may become cheaper if the compiler is able to do scalar replacement of the fields of the stack-allocated object
+(i.e., if the fields can be promoted).
+* Some field zero-initializations may be elided by the compiler.
+
+Object stack allocation is implemented in
+various Java runtimes.  This optimization is more important for Java since it doesn't have value types.
+
+## GitHub issues
+
+[roslyn #2104](https://github.com/dotnet/roslyn/issues/2104) Complier should optimize "alloc temporary small object" to "alloc on stack"
+
+[coreclr #1784](https://github.com/dotnet/coreclr/issues/1784) CLR/JIT should optimize "alloc temporary small object" to "alloc on stack" automatically
+
+## Escape Analysis
+
+An object is said to escape a method if it can be accessed after the method's execution has finished.
+An object allocation can be moved to the stack safely only if the object doesn't escape the allocating method.
+
+Several escape algorithms have been implemented in different Java implementations. Of the 3 algorithms listed in [references](References),
+[[1]](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.73.4799&rep=rep1&type=pdf)
+is the most precise and most expensive (it is based on connection graphs) and was used in the context of a static Java compiler,
+[[3]](https://pdfs.semanticscholar.org/1b33/dff471644f309392049c2791bca9a7f3b19c.pdf)
+is the least precise and cheapest (it doesn't track references through assignments of fields) and was used in MSR's Marmot implementation
+[[2]](https://www.usenix.org/legacy/events/vee05/full_papers/p111-kotzmann.pdf)
+is between
+[[1]](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.73.4799&rep=rep1&type=pdf) and 
+[[3]](https://pdfs.semanticscholar.org/1b33/dff471644f309392049c2791bca9a7f3b19c.pdf)
+both in analysis precision and cost. It was used in Java HotSpot.
+
+Effectiveness of object stack allocation depends in large part on whether escape analysis is done inter-procedurally.
+With intra-procedural analysis only, the compiler has to assume that arguments escape at all non-inlined call sites,
+which blocks many stack allocations. In particular, assuming that 'this' argument always escapes hurts the optimization.
+
```

Comment:
You might also mention that some approaches are able to  handle objects that only escape on some paths by promoting them to the heap "just in time" as control reaches those paths -- ffor instance [Partial Escape Analysis and Scalar Replacement for Java](https://dl.acm.org/citation.cfm?id=2544157).

So long as those escaping paths are indeed rare this can pay off.  For instance, in the local delegate case there is an exception path that makes it look like the delegate can escape. In my prototype I managed to modify the importer to prove this code wasn't reachable, but in general we may see things like this....

Reply:
Updated the doc to mention this approach and added the paper to References.